### PR TITLE
Verify and fix the `U256::mul_u64` method

### PR DESCRIPTION
**Pull request by tcharding**
Add kani verification for `U256::mul_u64`, doing so uncovered a bug in the current implementation due to overflow. Re-write the `mul_u64` method.

Fix: #1497

## Note

This PR now _only_ tests `mul_u64`, I will add testing div as a separate PR.

### CODE REVIEW

**Comment by Kixunil** [comment]
I think it'd be more useful if you also checked `(x * y) / y == x`

**Comment by Kixunil** [comment]
Oh, it found a bug, that's kinda cool!

**Comment by tcharding** [comment]
Please ignore all the pushes, I was on the wrong branch.

**Comment by tcharding** [comment]
Force push is total re-write, PR description has been updated.

**Comment by tcharding** [comment]
I don't rekon we should be running kani on every PR - it takes too long.

**Comment by elichai** [comment]
Can I propose this implementation instead? It's very straight forward:
```rust
    fn mul_u64(self, rhs: u64) -> (U256, bool) {
        let mut carry: u128 = 0;
        let mut splitted_le = [self.1 as u64, (self.1 >> 64) as u64, self.0 as u64, (self.0 >> 64) as u64];
        for word in &mut splitted_le {
            // TODO: Use `carrying_mul` when stabilized: https://github.com/rust-lang/rust/issues/85532
            let n = carry + (rhs as u128) * (*word as u128);
            *word = n as u64;
            carry = (n >> 64) & u64::MAX as u128;
        }
        let low = splitted_le[0] as u128 | (splitted_le[1] as u128) << 64;
        let high = splitted_le[2] as u128 | (splitted_le[3] as u128) << 64;
        (Self(high, low), carry != 0)
    }
```

**Comment by Kixunil** [comment]
Yeah, @elichai's impl looks better, ACK that one but I want all non-truncated casts to be changed to use `u128::from()` and add a comment that all truncating casts are intentional and correct. I've also manually iterated the algorithm for largest possible value to check that `+` doesn't overflow, so maybe add a comment about that too.

**Comment by elichai** [comment]
> Yeah, @elichai's impl looks better, ACK that one but I want all non-truncated casts to be changed to use `u128::from()` and add a comment that all truncating casts are intentional and correct. I've also manually iterated the algorithm for largest possible value to check that `+` doesn't overflow, so maybe add a comment about that too.

It can't overflow because 2 bitwise multiplications of `b` bits can never reach `2^b-1`. (for intuition, let's think of base 10, it's obvious there are no single-digit numbers that when multiplied can reach 99, you can also look at `u3` where `u3::MAX=7(0b111)` and `7*7 = 49(0b110001)` which is less than `u6::MAX=63=0b111111`))

**Comment by Kixunil** [comment]
> It can't overflow because 2 bitwise multiplications of `b` bits can never reach `2^b-1`

I think you meant to say "2 bitwise multiplications of `b` bits can never reach `2^(2*b)-1`", anyway this part was immediately obvious to me. The less obvious part is 2 multiplications with previous carry added not overflowing the same bound.

**Comment by elichai** [comment]
> > It can't overflow because 2 bitwise multiplications of `b` bits can never reach `2^b-1`
> 
> I think you meant to say "2 bitwise multiplications of `b` bits can never reach `2^(2*b)-1`", anyway this part was immediately obvious to me. The less obvious part is 2 multiplications with previous carry added not overflowing the same bound.

Right, I'm sorry, I'll try again:

∀ a,b ∈ ℤ  ∩ [0, 2^b - 1] : a*b ≤ (2^b-1)^2 = 2^{2b} - 2^{b+1} + 1
and because an integer with 2b bits is in the range `ℤ  ∩ [0, 2^{2b}-1]` then we have `2^{2b}-1 - (2^{2b} - 2^{b+1} + 1) = 2^{b+1} - 2` free space, and ∀ b ∈ ℤ, b > 1 we get `2^b ≤ 2^{b+1} - 2 `

**Comment by Kixunil** [comment]
OK, I guess a comment linking to your very nice proof is even better then. :) Thanks!

**Comment by elichai** [comment]
> OK, I guess a comment linking to your very nice proof is even better then. :) Thanks!

Doing a math degree does come in handy here and there 😜

**Comment by tcharding** [comment]
Thanks @elichai, mad geek props to you!

**Comment by tcharding** [comment]
Needs #1508 to pass CI.

**Review by elichai — DISMISSED** [review]
ACK

File: bitcoin/src/pow.rs [hunk 1]
```diff
@@ -421,32 +421,26 @@ impl U256 {
...
+    /// occurred. If an overflow occurred then the wrapped value is returned.
     fn mul_u64(self, rhs: u64) -> (U256, bool) {
-        // Multiply 64 bit parts of `mul` by `rhs`.
-        fn mul_parts(mul: u128, rhs: u64) -> (u128, u128) {
-            let upper = (rhs as u128) * (mul >> 64);
-            let lower = (rhs as u128) * (mul & 0xFFFF_FFFF_FFFF_FFFF);
-            (upper, lower)
-        }
+        let mut carry: u128 = 0;
+        let mut split_le = [self.1 as u64, (self.1 >> 64) as u64, self.0 as u64, (self.0 >> 64) as u64];
```
Comment (Kixunil) [inline bitcoin/src/pow.rs:430]: This is missing a comment about intentional truncation but maybe it's obvious enough from variable name.

File: bitcoin/src/pow.rs [hunk 2]
```diff
@@ -1613,3 +1607,29 @@ mod tests {
...
+    }
+
+    #[cfg(kani)]
+    #[kani::proof]
+    fn check_mul_u64() {
+        let high: u128 = kani::any();
+        let low: u128 = kani::any();
+        let y: u64 = kani::any();
+
+        let _ = U256(high, low).mul_u64(y);
```
Comment (Kixunil) [inline bitcoin/src/pow.rs:1633]: If there are more `kani` tests in the future we should impl `Arbitrary` for `U256`.

**Review by Kixunil — APPROVED** [review]
ACK fd35efd796a1452eabad6417d08ff766a92e2e5f

File: bitcoin/src/pow.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/pow.rs:430]: Yeah I tried using `buf` as the name and adding a comment but I think its clean like it is.

**Comment by tcharding** [comment]
Implemented `kani::Arbitrary` for `U256` - there is some dark magic going on that I don't understand, implementing a type from a crate that we do not depend on. Does `cargo kani` modify the manifest before building?

**Comment by tcharding** [comment]
Another question, kani CI failed but I get only green when running locally. `kani::any()` is non-deterministic but the final proof is deterministic, right?

**Comment by tcharding** [comment]
I don't know why kani is failing here?

**Comment by Kixunil** [comment]
> Does `cargo kani` modify the manifest before building?

Not sure if that or it calls rustc directly. But they do document that they inject the crate.

File: bitcoin/src/pow.rs (hunk 1 above)
Comment (Kixunil) [reply bitcoin/src/pow.rs:430]: I find `split_le` more clear than `buf`.

**Review by Kixunil — APPROVED** [review]
ACK cf9733d67886eff84622de4262863ec02827cb1f

**Review by apoelstra — APPROVED** [review]
ACK cf9733d67886eff84622de4262863ec02827cb1f

**Pull request merged**